File: app/src/main/java/com/close/hook/ads/ui/adapter/AppsAdapter.java
package com.close.hook.ads.ui.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;
import com.close.hook.ads.R;
import com.close.hook.ads.data.model.AppInfo;
import com.close.hook.ads.databinding.InstallsItemAppBinding;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.subjects.PublishSubject;

public class AppsAdapter extends ListAdapter<AppInfo, AppsAdapter.AppViewHolder> {

    private final PublishSubject<String> onClickSubject = PublishSubject.create();
    private final PublishSubject<String> onLongClickSubject = PublishSubject.create();

    private static final DiffUtil.ItemCallback<AppInfo> DIFF_CALLBACK = new DiffUtil.ItemCallback<AppInfo>() {
        @Override
        public boolean areItemsTheSame(@NonNull AppInfo oldItem, @NonNull AppInfo newItem) {
            return oldItem.getPackageName().equals(newItem.getPackageName());
        }

        @Override
        public boolean areContentsTheSame(@NonNull AppInfo oldItem, @NonNull AppInfo newItem) {
            boolean appNameSame = oldItem.getAppName() != null ? oldItem.getAppName().equals(newItem.getAppName()) : newItem.getAppName() == null;
            boolean versionNameSame = oldItem.getVersionName() != null ? oldItem.getVersionName().equals(newItem.getVersionName()) : newItem.getVersionName() == null;
            boolean versionCodeSame = oldItem.getVersionCode() == newItem.getVersionCode();

            return appNameSame && versionNameSame && versionCodeSame;
        }
    };

    public AppsAdapter() {
        super(DIFF_CALLBACK);
    }

    public Observable<String> getOnClickObservable() {
        return onClickSubject;
    }

    public Observable<String> getOnLongClickObservable() {
        return onLongClickSubject;
    }

    @NonNull
    @Override
    public AppViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        InstallsItemAppBinding binding = InstallsItemAppBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new AppViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull AppViewHolder holder, int position) {
        AppInfo appInfo = getItem(position);
        holder.bind(appInfo);

        holder.binding.getRoot().setOnClickListener(v -> onClickSubject.onNext(appInfo.getPackageName()));
        holder.binding.getRoot().setOnLongClickListener(v -> {
            onLongClickSubject.onNext(appInfo.getPackageName());
            return true;
        });
    }

    static class AppViewHolder extends RecyclerView.ViewHolder {
        final InstallsItemAppBinding binding;

        AppViewHolder(InstallsItemAppBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(AppInfo appInfo) {
            binding.appName.setText(appInfo.getAppName());
            binding.packageName.setText(appInfo.getPackageName());
            binding.appVersion.setText(appInfo.getVersionName() + " (" + appInfo.getVersionCode() + ")");
            if (appInfo.getAppIcon() != null) {
                Glide.with(binding.appIcon.getContext())
                        .load(appInfo.getAppIcon())
                        .apply(new RequestOptions()
                                .override(binding.appIcon.getContext().getResources().getDimensionPixelSize(R.dimen.app_icon_size))
                                .diskCacheStrategy(DiskCacheStrategy.RESOURCE))
                        .into(binding.appIcon);
            } else {
                binding.appIcon.setImageDrawable(null);
            }
        }
    }

    @Override
    public void onViewRecycled(@NonNull AppViewHolder holder) {
        super.onViewRecycled(holder);
        Glide.with(holder.binding.appIcon.getContext()).clear(holder.binding.appIcon);
    }
}

File: app/src/main/java/com/close/hook/ads/ui/adapter/AppsAdapter.kt
package com.close.hook.ads.ui.adapter

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.close.hook.ads.data.model.AppInfo
import com.close.hook.ads.databinding.InstallsItemAppBinding
import com.close.hook.ads.util.AppIconLoader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class AppsAdapter(
    private val onItemClickListener: OnItemClickListener
) : ListAdapter<AppInfo, AppsAdapter.AppViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppViewHolder {
        val binding = InstallsItemAppBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return AppViewHolder(binding, onItemClickListener)
    }

    override fun onBindViewHolder(holder: AppViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    override fun onViewRecycled(holder: AppViewHolder) {
        super.onViewRecycled(holder)
    }

    class AppViewHolder(
        private val binding: InstallsItemAppBinding,
        private val onItemClickListener: OnItemClickListener
    ) : RecyclerView.ViewHolder(binding.root) {

        private val targetSizePx by lazy { AppIconLoader.calculateTargetIconSizePx(binding.root.context) }

        init {
            binding.root.setOnClickListener {
                (binding.root.tag as? AppInfo)?.let { app ->
                    onItemClickListener.onItemClick(app, binding.appIcon.drawable)
                }
            }
            binding.root.setOnLongClickListener {
                (binding.root.tag as? AppInfo)?.let { app ->
                    onItemClickListener.onItemLongClick(app, binding.appIcon.drawable)
                }
                true
            }
        }

        fun bind(appInfo: AppInfo) {
            binding.root.tag = appInfo
            binding.appName.text = appInfo.appName
            binding.packageName.text = appInfo.packageName
            binding.appVersion.text = "${appInfo.versionName} (${appInfo.versionCode})"

            val context = binding.root.context
            val packageName = appInfo.packageName

            (context as? LifecycleOwner)?.lifecycleScope?.launch {
                val icon = AppIconLoader.loadAndCompressIcon(context, packageName, targetSizePx)

                if (binding.root.tag == appInfo) {
                    withContext(Dispatchers.Main) {
                        binding.appIcon.setImageDrawable(icon)
                    }
                }
            }
        }
    }

    interface OnItemClickListener {
        fun onItemClick(appInfo: AppInfo, icon: Drawable?)
        fun onItemLongClick(appInfo: AppInfo, icon: Drawable?)
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<AppInfo>() {
            override fun areItemsTheSame(oldItem: AppInfo, newItem: AppInfo): Boolean =
                oldItem.packageName == newItem.packageName
            override fun areContentsTheSame(oldItem: AppInfo, newItem: AppInfo): Boolean =
                oldItem == newItem
        }
    }
}
